tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetCupRule.scala

```diff
@@ -31,21 +31,35 @@ class SetCupRule(rewriter: SymbStateRewriter) extends RewritingRule {
         val leftSetCell = nextState.asCell
         nextState = rewriter.rewriteUntilDone(nextState.setRex(rightSet))
         val rightSetCell = nextState.asCell
-        val leftElems = nextState.arena.getHasPtr(leftSetCell)
-        val rightElems = nextState.arena.getHasPtr(rightSetCell)
+        val leftPtrs = nextState.arena.getHasPtr(leftSetCell)
+        val rightPtrs = nextState.arena.getHasPtr(rightSetCell)
 
-        val common = Set(leftElems: _*).intersect(Set(rightElems: _*))
-        val onlyLeft = Set(leftElems: _*).diff(common)
-        val onlyRight = Set(rightElems: _*).diff(common)
-        val allDistinct = common.toSeq ++ onlyLeft.toSeq ++ onlyRight.toSeq
+        val leftElems = leftPtrs.map(_.elem).toSet
+        val rightElems = rightPtrs.map(_.elem).toSet
+        val common = leftElems.intersect(rightElems)
+
+        val cellMap = (leftPtrs ++ rightPtrs).foldLeft(Map.empty[ArenaCell, Seq[ElemPtr]]) { case (m, ptr) =>
+          val elem = ptr.elem
+          m + (elem -> (m.getOrElse(elem, Seq.empty) :+ ptr))
+        }
```

Comment:
please add a one-line comment explaining what this code is doing. Is it building equivalence classes among pointers?